chain/client/src/view_client.rs

```diff
@@ -109,17 +110,26 @@ impl ViewClientActor {
     }
 
     fn handle_query(&mut self, msg: Query) -> Result<Option<QueryResponse>, String> {
-        if let Some(response) = self.query_responses.cache_remove(&msg.id) {
-            self.query_requests.cache_remove(&msg.id);
+        if let Some(response) = self.query_responses.cache_remove(&msg.query_id) {
+            self.query_requests.cache_remove(&msg.query_id);
             return response.map(Some);
         }
-        let header = self.chain.head_header().map_err(|e| e.to_string())?.clone();
-        let path_parts: Vec<&str> = msg.path.split('/').collect();
-        if path_parts.len() <= 1 {
-            return Err("Not enough query parameters provided".to_string());
-        }
-        let account_id = AccountId::from(path_parts[1].clone());
-        let shard_id = self.runtime_adapter.account_id_to_shard_id(&account_id);
+
+        let header = match msg.block_id {
+            Some(BlockId::Height(block_height)) => self.chain.get_header_by_height(block_height),
+            Some(BlockId::Hash(block_hash)) => self.chain.get_block_header(&block_hash),
+            None => self.chain.head_header(),
+        };
+        let header = header.map_err(|e| e.to_string())?.clone();
+
+        let account_id = match &msg.request {
+            QueryRequest::ViewAccount { account_id, .. } => account_id,
+            QueryRequest::ViewState { account_id, .. } => account_id,
+            QueryRequest::ViewAccessKey { account_id, .. } => account_id,
+            QueryRequest::ViewAccessKeyList { account_id, .. } => account_id,
+            QueryRequest::CallFunction { account_id, .. } => account_id,
+        };
+        let shard_id = self.runtime_adapter.account_id_to_shard_id(account_id);
```

Comment:
Do you need to check account ID validity before converting it to shard here?

Reply:
Well, I would like to improve `AccountId` type, so it enforces the contract of being a valid account ID. Let's make it in another PR; created an issue for that: https://github.com/nearprotocol/nearcore/issues/2074